src/PerfCounters_aarch64.h

```diff
@@ -1,35 +1,180 @@
 /* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
 // This file is included from PerfCounters.cc
 
-static const char* midr_path =
-    "/sys/devices/system/cpu/cpu0/regs/identification/midr_el1";
+struct CPUID {
+  uint8_t implementer{0};
+  uint8_t variant{0};
+  uint16_t part{0};
+  operator bool() const
+  {
+    return implementer || variant || part;
+  }
+  // bool operator==(const CPUID&) const = default; // c++20
+  bool operator==(const CPUID &other) const
+  {
+    return implementer == other.implementer &&
+      variant == other.variant && part == other.part;
+  }
+  bool operator!=(const CPUID &other) const
+  {
+    return !(*this == other);
+  }
+};
+static std::ostream &operator<<(std::ostream &stm, const CPUID &cpuid)
+{
+  stm << std::hex << "implementer: 0x" << int(cpuid.implementer)
+      << ", variant: 0x" << int(cpuid.variant) << " part: 0x" << int(cpuid.part);
+  return stm;
+}
 
 /**
  * Return the detected, known microarchitecture of this CPU, or don't
  * return; i.e. never return UnknownCpu.
  */
-static CpuMicroarch compute_cpu_microarch() {
-  FILE *midr_el1 = fopen(midr_path, "r");
-  if (!midr_el1) {
-    CLEAN_FATAL() << "Failed to read midr register from kernel";
-  }
-  uint32_t midir;
-  if (1 != fscanf(midr_el1, "%x", &midir)) {
-    CLEAN_FATAL() << "Failed to read midr register from kernel";
-  }
-  fclose(midr_el1);
-  switch (midir) {
-    case 0x413fd0c1:
+static CpuMicroarch compute_cpu_microarch(const CPUID &cpuid) {
+  switch (cpuid.implementer) {
+  case 0x41: // ARM
+    switch (cpuid.part) {
+    case 0xd0c:
       return ARMNeoverseN1;
-    default:
-      break;
+    }
+    break;
   }
-  CLEAN_FATAL() << "Aarch64 CPU type " << HEX(midir) << " unknown";
+  CLEAN_FATAL() << "Unknown aarch64 CPU type " << cpuid;
   return UnknownCpu; // not reached
 }
 
+static const std::string midr_dir = "/sys/devices/system/cpu/";
+static const std::regex cpuname_regex{"cpu([0-9]+)"};
```

Comment:
Here you can use parentheses instead of braces right?

Reply:
I think I'd rather have these inside `get_cpuinfo_sysfs`.

Reply:
Sure. (These are outside mainly because the old one used to, I'll move them in...)